java/adapter/avro/src/test/java/org/apache/arrow/AvroLogicalTypesTest.java

```diff
@@ -0,0 +1,168 @@
+    FieldVector vector = root.getFieldVectors().get(0);
+
+    checkPrimitiveResult(expected, vector);
+  }
+
+  @Test
+  public void testTimeMicros() throws Exception {
+    Schema schema = getSchema("logical/test_time_micros.avsc");
+
+    List<Long> data = Arrays.asList(10000L, 20000L, 30000L, 40000L, 50000L);
+
+    VectorSchemaRoot root = writeAndRead(schema, data);
+    FieldVector vector = root.getFieldVectors().get(0);
+
+    checkPrimitiveResult(data, vector);
+  }
+
+  @Test
+  public void testTimeMillis() throws Exception {
+    Schema schema = getSchema("logical/test_time_millis.avsc");
+
+    List<Integer> data = Arrays.asList(100, 200, 300, 400, 500);
+    List<LocalDateTime> expected = Arrays.asList(
+        DateUtility.getLocalDateTimeFromEpochMilli(100),
+        DateUtility.getLocalDateTimeFromEpochMilli(200),
+        DateUtility.getLocalDateTimeFromEpochMilli(300),
+        DateUtility.getLocalDateTimeFromEpochMilli(400),
+        DateUtility.getLocalDateTimeFromEpochMilli(500)
+    );
+
+    VectorSchemaRoot root = writeAndRead(schema, data);
+    FieldVector vector = root.getFieldVectors().get(0);
+
+    checkPrimitiveResult(expected, vector);
+  }
+
+  @Test
+  public void testDate() throws Exception {
+    Schema schema = getSchema("logical/test_date.avsc");
+
+    List<Integer> data = Arrays.asList(100, 200, 300, 400, 500);
+
+    VectorSchemaRoot root = writeAndRead(schema, data);
+    FieldVector vector = root.getFieldVectors().get(0);
+
+    checkPrimitiveResult(data, vector);
+  }
+
+  @Test
+  public void testDecimalWithOriginalBytes() throws Exception {
+    Schema schema = getSchema("logical/test_decimal_with_original_bytes.avsc");
+    List<ByteBuffer> data = Arrays.asList(
+        ByteBuffer.wrap(new BigDecimal("10").unscaledValue().toByteArray()),
+        ByteBuffer.wrap(new BigDecimal("1234").unscaledValue().toByteArray()),
+        ByteBuffer.wrap(new BigDecimal("123456").unscaledValue().toByteArray()),
+        ByteBuffer.wrap(new BigDecimal("111122").unscaledValue().toByteArray()));
+
+    List<BigDecimal> expected = Arrays.asList(
+        new BigDecimal(new BigInteger("10"), 2),
+        new BigDecimal(new BigInteger("1234"), 2),
+        new BigDecimal(new BigInteger("123456"), 2),
+        new BigDecimal(new BigInteger("111122"), 2));
+
+    VectorSchemaRoot root = writeAndRead(schema, data);
+    FieldVector vector = root.getFieldVectors().get(0);
+    checkPrimitiveResult(expected, vector);
+  }
+
+  @Test
+  public void testDecimalWithOriginalFixed() throws Exception {
+    Schema schema = getSchema("logical/test_decimal_with_original_fixed.avsc");
+
+    List<GenericData.Fixed> data = new ArrayList<>();
+    List<BigDecimal> expected = new ArrayList<>();
+
+    for (int i = 0; i < 5; i++) {
+      BigDecimal bigDecimal = new BigDecimal(i * i).setScale(2);
+      byte fillByte = (byte) (bigDecimal.signum() < 0 ? 0xFF : 0x00);
+      byte[] unscaled = bigDecimal.unscaledValue().toByteArray();
```

Comment:
this logic a little complex can you comment it/refactor it to helper methods or use existing functionality to make it clearer?